REGRESSION(ExtensionKit): Video stops playing on mlb.com when entering PiP

aestes opened the pull request:
#### 7223f04262f101f416e31198f62ca743ea803cdc
<pre>
REGRESSION(ExtensionKit): Video stops playing on mlb.com when entering PiP
<a href="https://bugs.webkit.org/show_bug.cgi?id=268532">https://bugs.webkit.org/show_bug.cgi?id=268532</a>
<a href="https://rdar.apple.com/121120961">rdar://121120961</a>

Reviewed by Per Arne Vollan.

Prior to managing WebKit auxiliary processes as extensions and enabling media capability grants,
the iOS media system tracked the visibility of the WebKit host process by observing the foreground
status of the app whose PID WebKit provided to AVSystemController via the
AVSystemController_PIDToInheritAppStateFrom SPI. When a PiP window is visible on screen then the
associated app and its WebContent and GPU processes are considered in the foreground and media
playback is allowed by the system.

Once WebKit started managing auxiliary processes as extensions we stopped using AVSystemController
SPIs and relied instead on BrowserEngineKit visibility propagation interactions to propagate
visibility from the host app to the WebContent and GPU extensions. Visibility propagation
interactions were installed on the WKContentView, and so long as the content view was in a visible
scene then the WebContent and GPU extensions would be considered visible by the media system.
However, the PiP window is in a different scene and may be visible at times when the
WKContentView&apos;s scene isn&apos;t. In these cases the WebContent and GPU extensions would not be
considered visible, preventing media playback from continuing in the PiP window while the host app
is in the backround.

Resolved this by encapsulating the logic for creating visibility propagation interactions into a
new UIView subclass (WKVisibilityPropagationView) that is added as a subview of WKContentViews and
WKLayerHostViews (the latter only when in a fullscreen or PiP presentation mode). This ensures that
WebContent and GPU extensions are considered visible whenever either of these views are in a visible
scene. WKContentView keeps a weak set of all WKVisibilityPropagationViews and is responsible for
creating and removing visibility propagation interactions as auxiliary processes launch and exit.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(-[WKLayerHostView visibilityPropagationView]):
(-[WKLayerHostView setVisibilityPropagationView:]):
(WebKit::VideoPresentationManagerProxy::setupFullscreenWithID):
(WebKit::VideoPresentationManagerProxy::didCleanupFullscreen):
(WebKit::VideoPresentationManagerProxy::setVideoLayerFrame):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::createVisibilityPropagationView):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::createVisibilityPropagationView):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
(-[WKContentView _installVisibilityPropagationViews]):
(-[WKContentView _setupVisibilityPropagationForWebProcess]):
(-[WKContentView _setupVisibilityPropagationForGPUProcess]):
(-[WKContentView _removeVisibilityPropagationViewForWebProcess]):
(-[WKContentView _removeVisibilityPropagationViewForGPUProcess]):
(-[WKContentView _didRelaunchProcess]):
(-[WKContentView _webProcessDidCreateContextForVisibilityPropagation]):
(-[WKContentView _gpuProcessDidCreateContextForVisibilityPropagation]):
(-[WKContentView _createVisibilityPropagationView]):
(-[WKContentView _setupVisibilityPropagationViewForWebProcess]): Deleted.
(-[WKContentView _setupVisibilityPropagationViewForGPUProcess]): Deleted.
* Source/WebKit/UIProcess/ios/WKVisibilityPropagationView.h: Added.
* Source/WebKit/UIProcess/ios/WKVisibilityPropagationView.mm: Added.
(-[WKVisibilityPropagationView propagateVisibilityToProcess:]):
(-[WKVisibilityPropagationView stopPropagatingVisibilityToProcess:]):
(-[WKVisibilityPropagationView _containsInteractionForProcess:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/274037@main">https://commits.webkit.org/274037@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d463caf20fb21aa361526911fedc156b62bc197

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37656 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16541 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39894 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40191 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33515 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13702 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38221 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13912 "Found 1 new test failure: media/video-ended-does-not-hold-sleep-assertion.html (failure)") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12156 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12091 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41450 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34009 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37986 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12679 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36155 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/14114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8472 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13079 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pvollan reviewed (approved):
R=me.

pvollan commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -815,25 +821,46 @@ - (void)_didRelaunchProcess
+    [self _setupVisibilityPropagationForGPUProcess];
 #endif
 
+    return visibilityPropagationView.autorelease();
```
Will the reference count of visibilityPropagationView be zero when the RetainPtr is deallocated upon returning from this function?

aestes replied:
No, `.autorelease()` prevents that.

pvollan reviewed (approved):
R=me.

aestes committed:
REGRESSION(ExtensionKit): Video stops playing on mlb.com when entering PiP
https://bugs.webkit.org/show_bug.cgi?id=268532
rdar://121120961

Reviewed by Per Arne Vollan.

Prior to managing WebKit auxiliary processes as extensions and enabling media capability grants,
the iOS media system tracked the visibility of the WebKit host process by observing the foreground
status of the app whose PID WebKit provided to AVSystemController via the
AVSystemController_PIDToInheritAppStateFrom SPI. When a PiP window is visible on screen then the
associated app and its WebContent and GPU processes are considered in the foreground and media
playback is allowed by the system.

Once WebKit started managing auxiliary processes as extensions we stopped using AVSystemController
SPIs and relied instead on BrowserEngineKit visibility propagation interactions to propagate
visibility from the host app to the WebContent and GPU extensions. Visibility propagation
interactions were installed on the WKContentView, and so long as the content view was in a visible
scene then the WebContent and GPU extensions would be considered visible by the media system.
However, the PiP window is in a different scene and may be visible at times when the
WKContentView's scene isn't. In these cases the WebContent and GPU extensions would not be
considered visible, preventing media playback from continuing in the PiP window while the host app
is in the backround.

Resolved this by encapsulating the logic for creating visibility propagation interactions into a
new UIView subclass (WKVisibilityPropagationView) that is added as a subview of WKContentViews and
WKLayerHostViews (the latter only when in a fullscreen or PiP presentation mode). This ensures that
WebContent and GPU extensions are considered visible whenever either of these views are in a visible
scene. WKContentView keeps a weak set of all WKVisibilityPropagationViews and is responsible for
creating and removing visibility propagation interactions as auxiliary processes launch and exit.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(-[WKLayerHostView visibilityPropagationView]):
(-[WKLayerHostView setVisibilityPropagationView:]):
(WebKit::VideoPresentationManagerProxy::setupFullscreenWithID):
(WebKit::VideoPresentationManagerProxy::didCleanupFullscreen):
(WebKit::VideoPresentationManagerProxy::setVideoLayerFrame):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::createVisibilityPropagationView):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::createVisibilityPropagationView):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
(-[WKContentView _installVisibilityPropagationViews]):
(-[WKContentView _setupVisibilityPropagationForWebProcess]):
(-[WKContentView _setupVisibilityPropagationForGPUProcess]):
(-[WKContentView _removeVisibilityPropagationViewForWebProcess]):
(-[WKContentView _removeVisibilityPropagationViewForGPUProcess]):
(-[WKContentView _didRelaunchProcess]):
(-[WKContentView _webProcessDidCreateContextForVisibilityPropagation]):
(-[WKContentView _gpuProcessDidCreateContextForVisibilityPropagation]):
(-[WKContentView _createVisibilityPropagationView]):
(-[WKContentView _setupVisibilityPropagationViewForWebProcess]): Deleted.
(-[WKContentView _setupVisibilityPropagationViewForGPUProcess]): Deleted.
* Source/WebKit/UIProcess/ios/WKVisibilityPropagationView.h: Added.
* Source/WebKit/UIProcess/ios/WKVisibilityPropagationView.mm: Added.
(-[WKVisibilityPropagationView propagateVisibilityToProcess:]):
(-[WKVisibilityPropagationView stopPropagatingVisibilityToProcess:]):
(-[WKVisibilityPropagationView _containsInteractionForProcess:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/274037@main
